kb/how-to/add-test-generation.md

```diff
@@ -0,0 +1,478 @@
+disabled.
+
+For the example, add the following class:
+
+```python
+from abc import ABCMeta
+...
+class BignumTarget(test_generation.BaseTarget, metaclass=ABCMeta):
+    #pylint: disable=abstract-method
+    """Target for bignum test case generation."""
+    # Using .ex in this example to avoid clash with `generate_bignum_tests.py`
+    target_basename = "test_suite_mpi.ex.generated"
+```
+
+Running the script will now create `test_suite_mpi.ex.generated.data`, and running with
+`--list` will list the filename as a valid target.
+
+### Adding the script to the build system
+
+This is not required for the example, but is included here for completeness.
+When adding a new script to the build system, the changes required include:
+ - Adding the script to
+[/tests/scripts/check-generated-files.sh](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/scripts/check-generated-files.sh).
+ - Adding the script to
+[/scripts/make_generated_files.bat](https://github.com/Mbed-TLS/mbedtls/blob/development/scripts/make_generated_files.bat).
+ - Adding calls and targets in CMake and Make for the script and its generated files.
+The script will need to be called in the same places where `generate_bignum_tests.py` is
+currently called.
+
+## Generating test cases for a function
+
+To add test cases for a function, a concrete class derived from `BaseTarget` must be
+added to the script.
+This class must implement all attributes and methods of `BaseTarget` and will generate
+test cases.
+This is a [test function class](#test-function-classes).
+
+As an example, test case generation is added for the test function `mpi_add_mpi()`.
+The added class will be derived from `BignumTarget`, and can be added either to
+[tests/scripts/generate_bignum_tests.py](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/scripts/generate_bignum_tests.py)
+or the example script described in the previous section.
+
+### Creating the class
+
+The class must derive from a file target class, and the following class attributes should
+be set:
+ - `count = 0`: this resets the test case counting mechanism.
+Alternatively, to disable test case numbering in descriptions, set
+`show_test_count = False`.
+ - `test_function`: the test function that the class will generate cases for.
+ - `test_name`: a short descriptive name for the test, this can be the same as
+`test_function` or a more readable equivalent.
+This is used as the start of the description line.
+
+For the example, add the following class:
+
+```python
+class BignumAddExample(BignumTarget):
+    """Test cases for bignum addition."""
+    count = 0
+    test_function = "mpi_add_mpi"
+    test_name = "MPI add"
+```
+
+### Class initialization
+
+Each instance of the class will represent a different test case, each with unique inputs.
+To initialize an instance of the class, these inputs must be passed to the `__init__()`
+constructor method.
+
+For `mpi_add_mpi()`, two input values are required, `val_a` and `val_b`.
```

Comment:
Almost everything in this paragraph is specific to generating bignum tests, and irrelevant otherwise. I think it would be more helpful to have a non-bignum or simplified example that doesn't go into these considerations about leading zeros, and a separate more advanced example that does illustrate how to deal with bignums. The more advanced example could be just a statement to go and look at a particular class in `generate_bignum_tests.py` though.